common/lib/xmodule/xmodule/docs/decisions/0001-move-due-dates-to-db.rst

```diff
@@ -0,0 +1,52 @@
+1. Move Due Dates to Relational Database
+----------------------------------------
+
+Status
+------
+
+Proposed
+
+Context
+-------
+
+We want an authoritative and readily-accessible place to track due dates for sections/units in a course.  We also want to override due dates per learner, and record the reason for the override. The current implementation stores due dates in the XBlock (in mongodb). If you want to know all of the due dates in a course, you have to walk the entire course structure and pull out each date, which is an inefficient operation.
+
+There is currently a way to override due dates per learner, using `IDDE <https://github.com/mitodl/ccx-idde-overrides-slides/blob/master/markdown/slides.md#individual-due-date-extensions-idde>`_, but this relies on a generic field-override system which doesn't know anything about the underlying data, and enabling this feature results in a database query for every xblock access, for every course on the platform. IDDE also does not allow for an audit trail on due date extensions, nor does it allow instructors to record a reason for the extension. 
+
+
+Decisions
+---------
+
+1. Create a new pluggable Django app responsible for dates and date overrides
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This approach anticipates other date-related functions in the future, such as APIs for returning a calendar view of upcoming dates.
+
+- The app will contain at least these models:
+    + ``ContentDate``: mapping content ids to dates
```

Comment:
I'm hoping we can get more normalized than this, FWIW. We have courses with literally a thousand problems in them, but those represent only dozens of assignments. I'm hoping we can derive those relationships and impose a little more structure on them.

Reply:
I think it might help if we could create an abstraction along the lines of "this is the policy around a scheduled thing, and that policy says it's due tomorrow", and make a mapping of content IDs to these policies -- without encoding either a strict sense of hierarchy or encoding due dates in an exhaustive set of nodes. As we move towards more dynamic content, I think having this level of indirection is going to be useful -- say you want to have multiple cohorts each see their own due dates for instance.